src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -25,24 +22,19 @@ namespace Microsoft.CodeAnalysis.Remote
     internal sealed partial class RemoteWorkspace : Workspace
     {
         /// <summary>
-        /// Guards updates to <see cref="_primaryBranchSolutionWithChecksum"/> and <see cref="_lastRequestedSolutionWithChecksum"/>.
+        /// Guards updates to all mutable state in this workspace.
         /// </summary>
-        private readonly SemaphoreSlim _availableSolutionsGate = new(initialCount: 1);
+        private readonly SemaphoreSlim _gate = new(initialCount: 1);
```

Comment:
note: our reads/writes of data should be tiny (and do not do async work while holding the gate).  So we could just consider a straight lock.  But i was slightly worried about N features kicking off and getting blocked behind a lock.